File: vane-portals/src/main/java/org/oddlama/vane/portals/PortalDynmapLayer.java
package org.oddlama.vane.portals;

import java.util.UUID;
import org.oddlama.vane.annotation.config.ConfigBoolean;
import org.oddlama.vane.annotation.config.ConfigInt;
import org.oddlama.vane.annotation.config.ConfigString;
import org.oddlama.vane.annotation.lang.LangMessage;
import org.oddlama.vane.core.lang.TranslatedMessage;
import org.oddlama.vane.core.module.Context;
import org.oddlama.vane.core.module.ModuleComponent;
import org.oddlama.vane.portals.portal.Portal;

public class PortalDynmapLayer extends ModuleComponent<Portals> {

    public static final String LAYER_ID = "vane_portals.portals";

    @ConfigInt(def = 29, min = 0, desc = "Layer ordering priority.")
    public int configLayerPriority;

    @ConfigBoolean(def = false, desc = "If the layer should be hidden by default.")
    public boolean configLayerHide;

    @ConfigString(def = "compass", desc = "The Dynmap marker icon.")
    public String configMarkerIcon;

    @LangMessage
    public TranslatedMessage langLayerLabel;

    @LangMessage
    public TranslatedMessage langMarkerLabel;

    private PortalDynmapLayerDelegate delegate = null;

    public PortalDynmapLayer(final Context<Portals> context) {
        super(
            context.group(
                "Dynmap",
                "Enable Dynmap integration. Public portals will then be shown on a separate Dynmap layer."
            )
        );
    }

    public void delayedOnEnable() {
        final var plugin = getModule().getServer().getPluginManager().getPlugin("Dynmap");
        if (plugin == null) {
            return;
        }

        delegate = new PortalDynmapLayerDelegate(this);
        delegate.onEnable(plugin);
    }

    @Override
    public void onEnable() {
        scheduleNextTick(this::delayedOnEnable);
    }

    @Override
    public void onDisable() {
        if (delegate != null) {
            delegate.onDisable();
            delegate = null;
        }
    }

    public void updateMarker(final Portal portal) {
        if (delegate != null) {
            delegate.updateMarker(portal);
        }
    }

    public void removeMarker(final UUID portalId) {
        if (delegate != null) {
            delegate.removeMarker(portalId);
        }
    }

    public void updateAllMarkers() {
        if (delegate != null) {
            delegate.updateAllMarkers();
        }
    }
}

File: vane-portals/src/main/kotlin/org/oddlama/vane/portals/PortalDynmapLayer.kt
package org.oddlama.vane.portals

import org.oddlama.vane.annotation.config.ConfigBoolean
import org.oddlama.vane.annotation.config.ConfigInt
import org.oddlama.vane.annotation.config.ConfigString
import org.oddlama.vane.annotation.lang.LangMessage
import org.oddlama.vane.core.lang.TranslatedMessage
import org.oddlama.vane.core.module.Context
import org.oddlama.vane.core.module.ModuleComponent
import org.oddlama.vane.portals.portal.Portal
import java.util.*

/**
 * Module component that integrates public portal markers with Dynmap.
 *
 * @property configLayerPriority display priority of the Dynmap layer.
 * @property configLayerHide whether the layer is hidden by default.
 * @property configMarkerIcon Dynmap icon id for portal markers.
 * @property langLayerLabel localized layer label.
 * @property langMarkerLabel localized marker label template.
 */

class PortalDynmapLayer(context: Context<Portals?>) : ModuleComponent<Portals?>(
    context.group(
        "Dynmap",
        "Enable Dynmap integration. Public portals will then be shown on a separate Dynmap layer."
    )
) {
    /** Display priority for the Dynmap layer. */
    @JvmField
    @ConfigInt(def = 29, min = 0, desc = "Layer ordering priority.")
    var configLayerPriority: Int = 0

    /** Whether the Dynmap layer is hidden by default. */
    @JvmField
    @ConfigBoolean(def = false, desc = "If the layer should be hidden by default.")
    var configLayerHide: Boolean = false

    /** Dynmap marker icon id for portal markers. */
    @JvmField
    @ConfigString(def = "compass", desc = "The Dynmap marker icon.")
    var configMarkerIcon: String? = null

    /** Localized Dynmap layer label. */
    @JvmField
    @LangMessage
    var langLayerLabel: TranslatedMessage? = null

    /** Localized marker label template. */
    @JvmField
    @LangMessage
    var langMarkerLabel: TranslatedMessage? = null

    /** Delegate that performs Dynmap API operations. */
    private var delegate: PortalDynmapLayerDelegate? = null

    /** Enables Dynmap integration after plugin startup has settled. */
    fun delayedOnEnable() {
        if (module!!.server.pluginManager.getPlugin("Dynmap") == null) return

        delegate = PortalDynmapLayerDelegate(this)
        delegate!!.onEnable()
    }

    /** Schedules deferred Dynmap integration setup. */
    public override fun onEnable() {
        scheduleNextTick { delayedOnEnable() }
    }

    /** Tears down Dynmap integration state. */
    public override fun onDisable() {
        delegate?.onDisable()
        delegate = null
    }

    /** Updates or creates the Dynmap marker for [portal]. */
    fun updateMarker(portal: Portal?) {
        portal?.let { delegate?.updateMarker(it) }
    }

    /** Removes the Dynmap marker for the given portal id. */
    fun removeMarker(portalId: UUID?) {
        portalId?.let { delegate?.removeMarker(it) }
    }

    /** Rebuilds all portal markers in Dynmap. */
    fun updateAllMarkers() {
        delegate?.updateAllMarkers()
    }

    /** Shared constants for Dynmap integration. */
    companion object {
        /** Marker set id used for the portal layer. */
        const val LAYER_ID: String = "vane_portals.portals"
    }
}
